### PR TITLE
NNX Migration: all commits

### PR DESCRIPTION
**Pull request by xibinliu**
# Description

Contains all the commits to make the NNX migrated MaxText work.

# Tests

Current focus is on:
- pre-training with nnx, with these flags: "pure_nnx_decoder=True enable_nnx=True pure_nnx=True fused_qkv=False"

Example:
```
python3  -m MaxText.train src/MaxText/configs/base.yml run_name="xibin-nnx" model_name="gpt3-52k"  dataset_type=synthetic steps=10 enable_checkpointing=False pure_nnx_decoder=True enable_nnx=True pure_nnx=True fused_qkv=False
```

More features and models will be tested and compared with Linen.
[Test track doc](https://docs.google.com/document/d/1NbUP3g5glgbC6bMyt44pwM_vQA1NR7U2rBUzfbTDwSs/edit?pli=1&resourcekey=0-9EUahtzL-hCycdu7l0grhQ&tab=t.rq5rwmyrc9bq)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/experimental/rl/grpo_trainer.py [hunk 1]
```diff
@@ -543,23 +543,43 @@ def setup_train_loop(
...
     inference_mesh = inference_model.mesh
-    init_rng, checkpoint_manager, learning_rate_schedule, tx = train_utils.create_training_tools(config, model, mesh)
+    init_rng = jax.random.PRNGKey(config.init_weights_seed)
+    learning_rate_schedule, tx = train_utils.create_training_optimizer(config, model)
+    if config.pure_nnx:
+      # NNX has a different function to init the training state.
+      raise NotImplementedError("Pure NNX support has not been implemented yet.")
+    else:
+      init_state_fn = functools.partial(maxtext_utils.init_initial_state, model, tx, config, True, init_rng)
+    checkpoint_manager = train_utils.create_checkpoint_manager(config, mesh, init_state_fn)
```
Comment (charlesli640) [inline src/maxtext/experimental/rl/grpo_trainer.py]: Need follow latest code.
